Usman_Assignments/recursion.py

```diff
@@ -0,0 +1,165 @@
+from myarray import MyArray
+from doubly_linkedlist import LinkedList
+
+# class Node(object):
+#     def __init__(self, data = None, next = None, prev = None):
+#         super(Node, self).__init__()
+#         self.data = data
+#         self.next = next
+#         self.prev = prev
+
+#     def __str__(self):
+#         return "{} -> {}".format(self.data, self.next)
+
+class Recursion (object):
+    def __init__ (self):
+        self.array = MyArray('i', 1, 2, 3, 4, 5)
+        self.sum = 0
+        self.linkedlist = LinkedList()
+
+    # input: n, any non-negative integer
+    def print_0toN (self, n):
+        if n < 0:
+            raise ValueError("Negative input number")
+            return
+
+        # 0, 1, 2, 3, 4, 5
+        # (5) => (4) => (3) => (2) => (1) => (0) <---- base case (0)
+        #  5  <=  4  <=  3  <=  2  <=  1  <=  0 
+
+        # base case
+        if n == 0:
+            print(n)
+            return
+
+        else:
+            # Hypothesis
+            self.print_0toN(n - 1)
+            # Induction
+            print(n)
+            return
+
+    def print_negative_NtoN(self, n):
+
+    #       -2, -1, 0, 1, 2
+    # =================================
+    # (2) => (1) => (0) => (-1) => (-2)
+    # -2  <= -1  <=  0  <=   1  <=   2
+
+        if N == 0:
+            print(n)
+            return
+
+        # base case
+        if N == -n:
+            if n == 0:
+                self.print_negative_NtoN(n - 1)
+            
+            print (n)
+            return
+
+        else:
+            # Hypothesis
+            self.print_negative_NtoN(n - 1)
+            # Induction
+            print(n)
+            return
+
+    
+
+    # print ("print_arr function is running")
+    def print_arr (self, arr, start, len):
+        
+        # base case
+        if (start >= len):
+            return
+
+        print (self.array.get(start))
+        self.print_arr (self.array, start + 1, len)
+        return
+
+    def print_reverse_arr(self, arr, end, start):
+        # base case
+        if (end < start):
+            return
+        
+        print (arr.get(end))
+        self.print_reverse_arr(arr, end - 1, start)
+        return
+
+    def sumOfArray(self, arr, start):
+        if (start >= arr.size()):
+            return
+
+        self.sum += arr.get(start)
+        self.sumOfArray(arr, start + 1)
+        return self.sum
+
+    def print_LinkedList(self, node):
+        if node is None:
+            return
+
+        print (node.data)
+        self.print_LinkedList(node.next)
+        return
+
+    def print_reverse_LinkedList(self, node):
+        if node is None:
+            return
+
+        print (node.data)
+        self.print_reverse_LinkedList(node.prev)

```

Comment:
How to do it given head pointer to print in reverse order instead of tail pointer?